<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>DuctApe</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="Analyze and link genomics and phenomics experiments">
    <meta name="author" content="Marco Galardini" >
    
    <!-- Le meta tag to refresh the cache -->
    <meta http-equiv="last-modified" content="Fri, 12 Oct 2012 11:00:00 GMT" />

    <!-- Le styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
	<link href="css/custom.css" rel="stylesheet">
    <style type="text/css">
      body {
        padding-top: 60px;
        padding-bottom: 40px;
      }
    </style>

    <!-- Le HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="//html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le fav and touch icons -->
    <link rel="shortcut icon" href="favicon.ico">
  </head>

  <body>

    <div class="container">
      <div id="home">
        <!--<div class="hero-unit">-->
		<div class="row">
			<noscript>
			  <div class="alert alert-error">
				<strong>Whoops! You have javascript disabled</strong>
				<br/>                
				You may be unable to enjoy this site, sorry :(
			  </div>
			 </noscript>
			 
			<div class="span6">
				<div align="center">
					<img alt="The almighty gorilla" src="img/logo.png" />
					<br>
					<img src="img/scotch.png" width="350px"/>
				</div>
			</div>
			
			<br>

			<div class="span6">
				<div class="hero-unit">
					<h3>Analyze and link genomics and phenomics experiments</h3>
					<p>
						DuctApe is a python library with a series of command-line programs that will help
						bioinformaticians to analyze genomes AND phenomic experiments. <br> The final purpose
						of the program is to combine the genomic informations (encoded as KEGG pathways)
						with the results of phenomic experiments (for now the Phenotype microarrays are supported) 
						and highlight the genes that may be responsible for phenotypic variations.
					</p>
					<br>
					<a href="https://github.com/downloads/combogenomics/DuctApe/DuctApe-0.3.4.tar.gz" class="btn btn-huge">
					<i class="icon-download"></i> Download &raquo;</a>
					<a href="https://github.com/combogenomics/DuctApe" class="btn btn-huge">
					<i class="icon-eye-open"></i> Look at the code &raquo;</a>
				</div>
			</div>
			 
		</div>
		
		<br>
		
		<div class="row">
			<div class="span4">
				<div class="hero-unit">
					<h3>Features</h3>
					<dl>
						<dt>Powerful</dt>
							<dd>Many kind of experiment are supported:
							single organism, mutants, pangenome</dd>
						<dt>Flexible</dt>
							<dd>Three command-line tools with git-like syntax, to allow maximum 
							flexibility in the analysis</dd>
						<dt>Modular</dt>
							<dd>Each component of DuctApe is encapsulated in a thread, to simplify further 
							developments of client applications (i.e. GUIs or web interfaces)</dd>
						<dt>Portable</dt>
							<dd>All the analysis are saved in a single file (a SQLite database), 
							to avoid complexity and to encourage exchange of data</dd>
					</dl>
				</div>
			</div>
			
			<div class="span8">
				<div class="hero-unit">
					<h3>The three modules of DuctApe</h3>
					<p>
						<ul>
							<h3>
								<li>
									dape <img src="img/scotch.png" height="60px"/>
								</li>
							</h3>
							
							<ul>
								<li>Used to initialize the project and add/remove the organism</li>
								<li>Performs the combination of genomics and phenomics experiments (*)</li>
							</ul>
							
							<br>
							
							<h3>
								<li>
									dgenome <img src="img/scotchgenome.png" height="60px"/>
								</li>
							</h3>
							
							<ul>
								<li>Handles protein sequences for each organism of the project</li>
								<li>For pangenomic studies it can build the pangenome using 
								the BBH algorithm (user-defined pangenomes can be added)</li>
								<li>The proteins are mapped to KEGG metabolic pathways</li>
							</ul>
							
							<br>
							
							<h3>
								<li>
									dphenome <img src="img/scotchphenome.png" height="60px"/>
								</li>
							</h3>
							
							<ul>
								<li>Handles phenomic experiments for each organism of the project (Phenotype microarray experiments)</li>
								<li>Performs zero-subtraction on the biolog plates</li>
								<li>Calculates the growth parameters</li>
								<li>Ranks the single experiments with a clusterization over growth parameters</li>
								<li>The compounds from the phenomic experiments are mapped to KEGG</li>
							</ul>
						</ul>
						
					</p>
					
					<br>
						
					<div class="bio">
						(*) not implemented yet
					</div>
				</div>
			</div>
		</div>
		
		<div class="row">
			<div class="span4">
				<div class="hero-unit">
					<h3>Brought to you by</h3>
					<p>
						<p>
							<div class="bio">
								<img align="left" src="img/marco.png" width="50px"><b> Marco Galardini</b>
								<br>
								Phd student @ <a href="http://www.unifi.it/">unifi</a>
								<br>
								marco.galardini AT unifi DOT it
							</div>
						</p>
						
						<div class="bio">
							<a href="http://www.unifi.it/dbefcb/"><img src="img/combo.png" width="100px"/></a>
							DuctApe is a <a href="http://www.unifi.it/dbefcb/">Combo</a> project
						</div>
						
					</p>
				</div>
			</div>
			
			<div class="span8">
				<div class="hero-unit">
					<h3>Quick how-to</h3>
					
						<br>
						
						<h3>Requirements</h3>
						<div class="bio">
							DuctApe has a few python library dependencies:
						</div>
						<ul>
							<li>Biopython</li>
							<li>NumPy</li>
							<li>SciPy</li>
							<li>matplotlib</li>
							<li>scikits.learn</li>
							<li>SOAPpy</li>
							<li>multiprocessing</li>
							<li>networkx (*)</li>
						</ul>
						
						<br>
						
						<h3>Installation</h3>
							<ul>
								<li>Method 1 (dependencies won't be checked)</li>
									<code>sudo python setup.py install</code>
								
								<li>Method 2 (missing dependencies will be downloaded)</li>
									<code>sudo pip install DuctApe-X.X.X.tar.gz</code>
								
							</ul>
							
						<br>
						
						<h3>Usage examples</h3>
							<ol>
								<h4><li>Single organism experiment</li></h4>
								<pre># Initialize the project
dape init
# add my organism using the ID MyOrg
dape add MyOrg

# add the proteome of MyOrg
dgenome add MyOrg.faa MyOrg
# adds the output of KAAS, a KEGG mapper
dgenome add-ko MyOrg.tab
# map the proteome to KEGG
dgenome start
# output the KEGG metabolic maps
dgenome map
# statistics and graphics 
dgenome stats
# export the genomic data
dgenome export

# add the phenomic experiment, BIOLOG data
dphenome add MyOrg.csv MyOrg
# perform control subtraction
dphenome zero
# calculate the growth parameters and perform the clusterization, using 4 CPUs
dphenome start -n 4
# plot the growth curves
dphenome plot
# remove inconsistent replicas: keep the highest replicas when there is an activity index delta >= 3
dphenome purge -d 3 keep-max
# plot only those curves that are not purged
dphenome plot
# restore the purged replicas
dphenome restore
dphenome stats
dphenome export

dape map (*)</pre>
								<br>

								<h4><li>Mutant experiment</li></h4>
								<pre>dape init
dape add MyOrg
# add mutant MyMut, a deletion mutant of MyOrg
dape add-mut -m MyOrg -k deletion MyMut

# add the proteins files found in this directory
dgenome add-dir MyFolder
dgenome add-ko MyOrg.tab
dgenome start
# plot only tha maps of the mutant
dgenome map MyMut
# add the phenomic files found in this directory
dphenome add-dir MyPhenomicFolder
dphenome zero
dphenome start -n 4
dphenome purge -d 3 keep-max
dphenome plot

dape map (*)</pre>

								<h4><li>Pangenomic experiment</li></h4>
								<pre>dape init
dape add MyOrg
dape add MyOrg2
dape add MyOrg3

dgenome add-dir MyFolder
dgenome add-ko MyOrg.tab MyOrg2.tab MyOrg3.tab
# also perform pangenome creation using 4 CPUs
dgenome start -n 4
# plot the maps for the whole pangenome
dgenome map

dphenome add-dir MyPhenomicFolder
dphenome zero
dphenome start -n 4
dphenome purge -d 3 keep-max
dphenome plot

dape map (*)</pre>
							</ol>
							
							<br>
						
					<div class="bio">
						(*) not implemented yet
					</div>
				</div>
			</div>
		</div>
		
      </div>
      
      <footer class="footer">
	  </footer>
    
    </div>

    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="js/jquery.js"></script>
    <script src="js/bootstrap.js"></script>
    
    <!-- Le Google analytics -->
    <script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-34201518-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>

  </body>
</html>
